### PR TITLE
Fixes #2751 - Commit suicide on ZK exceptions

### LINKED ISSUE #2751: Leader not aware of its leadership
Our cluster got into a state in which a leader got elected, but the elected leader wasn't aware of it's leadership, so it was refusing the requests the other leaders were proxying.

This state persisted for days until the "confused leader" was resterted.

Logs from the leader:

```
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,064] INFO Candidate /marathon/leader/member_0000000527 waiting for the next leader election, current voting: [member_0000000521, member_0000000522, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,064] ERROR Current member ID member_0000000492 is not a candidate for leader, current voting: [member_0000000521, member_0000000522, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,282] INFO Candidate /marathon/leader/member_0000000527 waiting for the next leader election, current voting: [member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,282] ERROR Current member ID member_0000000492 is not a candidate for leader, current voting: [member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,802] INFO Candidate /marathon/leader/member_0000000527 waiting for the next leader election, current voting: [member_0000000528, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,803] ERROR Current member ID member_0000000492 is not a candidate for leader, current voting: [member_0000000528, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,976] INFO Candidate /marathon/leader/member_0000000527 waiting for the next leader election, current voting: [member_0000000528, member_0000000529, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:22,976] ERROR Current member ID member_0000000492 is not a candidate for leader, current voting: [member_0000000528, member_0000000529, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:23 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:23,260] INFO Do not proxy to myself. Waiting for consistent leadership state. Are we leader?: false, leader: Some(srv2.hw.ca1.mesosphere.com:8080) (mesosphere.marathon.api.LeaderProxyFilter$:qtp983834017-106976)
Nov 30 13:51:23 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:23,261] INFO Waiting for consistent leadership state. Are we leader?: false, leader: Some(srv2.hw.ca1.mesosphere.com:8080) (mesosphere.marathon.api.LeaderProxyFilter$:qtp983834017-106976)
Nov 30 13:51:23 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:23,266] INFO Do not proxy to myself. Waiting for consistent leadership state. Are we leader?: false, leader: Some(srv2.hw.ca1.mesosphere.com:8080) (mesosphere.marathon.api.LeaderProxyFilter$:qtp983834017-107011)
Nov 30 13:51:23 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:23,267] INFO Waiting for consistent leadership state. Are we leader?: false, leader: Some(srv2.hw.ca1.mesosphere.com:8080) (mesosphere.marathon.api.LeaderProxyFilter$:qtp983834017-107011)
Nov 30 13:51:23 srv2.hw.ca1.mesosphere.com marathon[2379]: [2015-11-30 13:51:23,512] INFO Waiting for consistent leadership state. Are we leader?: false, leader: Some(srv2.hw.ca1.mesosphere.com:8080) (mesosphere.marathon.api.LeaderProxyFilter$:qtp983834017-106976)
```

Logs from another master, which apparently had gotten into the same state, but managed to get out of it:

```
Nov 30 13:51:19 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:19,122] INFO Waiting for consistent leadership state. Are we leader?: false, leader: Some(srv3.hw.ca1.mesosphere.com:8080) (mesosphere.marathon.api.LeaderProxyFilter$:qtp983834017-219000)
Nov 30 13:51:19 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:19,129] INFO Waiting for consistent leadership state. Are we leader?: false, leader: Some(srv3.hw.ca1.mesosphere.com:8080) (mesosphere.marathon.api.LeaderProxyFilter$:qtp983834017-219006)
Nov 30 13:51:19 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:19,366] INFO 172.16.0.14 - - [30/Nov/2015:13:51:16 +0000] "GET //srv3.hw.ca1.mesosphere.com:8080/v2/apps HTTP/1.1" 503 351 "http://ovh.marathon.io:4242/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36" 275 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-219011)
Nov 30 13:51:19 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:19,373] INFO 172.16.0.14 - - [30/Nov/2015:13:51:16 +0000] "GET //srv3.hw.ca1.mesosphere.com:8080/v2/deployments HTTP/1.1" 503 358 "http://ovh.marathon.io:4242/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36" 275 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-219000)
Nov 30 13:51:19 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:19,379] INFO 172.16.0.14 - - [30/Nov/2015:13:51:16 +0000] "GET //srv3.hw.ca1.mesosphere.com:8080/v2/queue HTTP/1.1" 503 352 "http://ovh.marathon.io:4242/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36" 275 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-219006)
Nov 30 13:51:21 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:21,702] INFO Unable to read additional data from server sessionid 0x3515255abc00003, likely server has closed socket, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(srv4.hw.ca1.mesosphere.com:2181))
Nov 30 13:51:21 srv3.hw.ca1.mesosphere.com marathon[2335]: 2015-11-30 13:51:21,704:2206(0x7f1a1cbe0700):ZOO_ERROR@handle_socket_error_msg@1721: Socket [172.16.0.14:2181] zk retcode=-4, errno=112(Host is down): failed while receiving a server response
Nov 30 13:51:21 srv3.hw.ca1.mesosphere.com marathon[2335]: 2015-11-30 13:51:21,704:2206(0x7f1a1cbe0700):ZOO_INFO@check_events@1703: initiated connection to server [172.16.0.15:2181]
Nov 30 13:51:21 srv3.hw.ca1.mesosphere.com marathon[2335]: I1130 13:51:21.704942 14894 group.cpp:418] Lost connection to ZooKeeper, attempting to reconnect ...
Nov 30 13:51:21 srv3.hw.ca1.mesosphere.com marathon[2335]: 2015-11-30 13:51:21,760:2206(0x7f1a1cbe0700):ZOO_INFO@check_events@1750: session establishment complete on server [172.16.0.15:2181], sessionId=0x15150291daa0005, negotiated timeout=10000
Nov 30 13:51:21 srv3.hw.ca1.mesosphere.com marathon[2335]: I1130 13:51:21.760956 14922 group.cpp:313] Group process (group(3)@172.16.0.13:50346) reconnected to ZooKeeper
Nov 30 13:51:21 srv3.hw.ca1.mesosphere.com marathon[2335]: I1130 13:51:21.760977 14922 group.cpp:787] Syncing group operations: queue size (joins, cancels, datas) = (0, 0, 0)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,402] INFO Opening socket connection to server srv2.hw.ca1.mesosphere.com/172.16.0.12:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(srv2.hw.ca1.mesosphere.com:2181))
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,402] INFO Socket connection established to srv2.hw.ca1.mesosphere.com/172.16.0.12:2181, initiating session (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(srv2.hw.ca1.mesosphere.com:2181))
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,550] INFO Unable to reconnect to ZooKeeper service, session 0x3515255abc00003 has expired, closing socket connection (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(srv2.hw.ca1.mesosphere.com:2181))
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,550] INFO Zookeeper session expired. Event: WatchedEvent state:Expired type:None path:null (com.twitter.common.zookeeper.ZooKeeperClient:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,550] INFO EventThread shut down (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,551] INFO Defeated (Leader Interface) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,551] INFO Call onDefeated leadership callbacks on EntityStoreCache(MarathonStore(app:)), EntityStoreCache(MarathonStore(group:)), EntityStoreCache(MarathonStore(deployment:)), EntityStoreCache(MarathonStore(framework:)), EntityStoreCache(MarathonStore(taskFailure:)), EntityStoreCache(MarathonStore(task:)), EntityStoreCache(MarathonStore(events:)) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuic
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,551] INFO Finished onDefeated leadership callbacks (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,552] INFO Defeat leadership (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,552] INFO Now standing by. Closing existing handles and rejecting new. (mesosphere.marathon.event.http.HttpEventStreamActor:marathon-akka.actor.default-dispatcher-75)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,553] INFO Stopping driver (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,553] INFO Initiating client connection, connectString=srv2.hw.ca1.mesosphere.com:2181,srv3.hw.ca1.mesosphere.com:2181,srv4.hw.ca1.mesosphere.com:2181,srv5.hw.ca1.mesosphere.com:2181,srv6.hw.ca1.mesosphere.com:2181 sessionTimeout=10000 watcher=com.twitter.common.zookeeper.ZooKeeperClient$3@2a4c6776 (org.apache.zookeeper.ZooKeeper:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,555] INFO Opening socket connection to server srv3.hw.ca1.mesosphere.com/172.16.0.13:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(srv3.hw.ca1.mesosphere.com:2181))
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,555] INFO Socket connection established to srv3.hw.ca1.mesosphere.com/172.16.0.13:2181, initiating session (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(srv3.hw.ca1.mesosphere.com:2181))
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,806] INFO Session establishment complete on server srv3.hw.ca1.mesosphere.com/172.16.0.13:2181, sessionid = 0x25152e0bae20017, negotiated timeout = 10000 (org.apache.zookeeper.ClientCnxn:ZookeeperClient-watcherProcessor-SendThread(srv3.hw.ca1.mesosphere.com:2181))
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,807] ERROR Current member ID member_0000000494 is not a candidate for leader, current voting: [member_0000000528, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,808] INFO Defeated (Leader Interface) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,808] INFO Call onDefeated leadership callbacks on EntityStoreCache(MarathonStore(app:)), EntityStoreCache(MarathonStore(group:)), EntityStoreCache(MarathonStore(deployment:)), EntityStoreCache(MarathonStore(framework:)), EntityStoreCache(MarathonStore(taskFailure:)), EntityStoreCache(MarathonStore(task:)), EntityStoreCache(MarathonStore(events:)) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuic
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,808] INFO Finished onDefeated leadership callbacks (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,808] INFO Defeat leadership (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,808] INFO Now standing by. Closing existing handles and rejecting new. (mesosphere.marathon.event.http.HttpEventStreamActor:marathon-akka.actor.default-dispatcher-73)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,809] INFO Stopping driver (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,810] ERROR Current member ID member_0000000506 is not a candidate for leader, current voting: [member_0000000528, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,810] INFO Defeated (Leader Interface) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,810] INFO Call onDefeated leadership callbacks on EntityStoreCache(MarathonStore(app:)), EntityStoreCache(MarathonStore(group:)), EntityStoreCache(MarathonStore(deployment:)), EntityStoreCache(MarathonStore(framework:)), EntityStoreCache(MarathonStore(taskFailure:)), EntityStoreCache(MarathonStore(task:)), EntityStoreCache(MarathonStore(events:)) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuic
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,810] INFO Finished onDefeated leadership callbacks (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,810] INFO Defeat leadership (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,810] INFO Now standing by. Closing existing handles and rejecting new. (mesosphere.marathon.event.http.HttpEventStreamActor:marathon-akka.actor.default-dispatcher-76)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,811] INFO Stopping driver (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,811] ERROR Current member ID member_0000000507 is not a candidate for leader, current voting: [member_0000000528, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,811] INFO Defeated (Leader Interface) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,811] INFO Call onDefeated leadership callbacks on EntityStoreCache(MarathonStore(app:)), EntityStoreCache(MarathonStore(group:)), EntityStoreCache(MarathonStore(deployment:)), EntityStoreCache(MarathonStore(framework:)), EntityStoreCache(MarathonStore(taskFailure:)), EntityStoreCache(MarathonStore(task:)), EntityStoreCache(MarathonStore(events:)) (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuic
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,811] INFO Finished onDefeated leadership callbacks (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,811] INFO Defeat leadership (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,812] INFO Now standing by. Closing existing handles and rejecting new. (mesosphere.marathon.event.http.HttpEventStreamActor:marathon-akka.actor.default-dispatcher-73)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,812] INFO Stopping driver (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$1bda83eb:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,980] INFO Set group member ID to member_0000000529 (com.twitter.common.zookeeper.Group:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,981] ERROR Current member ID member_0000000506 is not a candidate for leader, current voting: [member_0000000528, member_0000000529, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,981] ERROR Current member ID member_0000000494 is not a candidate for leader, current voting: [member_0000000528, member_0000000529, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,981] INFO Candidate /marathon/leader/member_0000000529 waiting for the next leader election, current voting: [member_0000000528, member_0000000529, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor)
Nov 30 13:51:22 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:51:22,982] ERROR Current member ID member_0000000507 is not a candidate for leader, current voting: [member_0000000528, member_0000000529, member_0000000525, member_0000000526, member_0000000527] (com.twitter.common.zookeeper.CandidateImpl:ZookeeperClient-watcherProcessor-EventThread)
Nov 30 13:52:01 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:01,693] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-219003)
Nov 30 13:52:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:04,255] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-219016)
Nov 30 13:52:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:04,383] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-218989)
Nov 30 13:52:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:04,458] INFO 0:0:0:0:0:0:0:1 - - [30/Nov/2015:13:52:01 +0000] "GET //localhost:8080/v2/tasks HTTP/1.1" 503 352 "-" "python-requests/2.6.0 CPython/2.7.5 Linux/3.10.0-229.7.2.el7.x86_64" 276 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-219003)
Nov 30 13:52:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:04,686] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-219012)
Nov 30 13:52:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:04,960] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-218990)
Nov 30 13:52:07 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:07,013] INFO 172.16.1.13 - - [30/Nov/2015:13:52:04 +0000] "GET //int.srv3.hw.ca1.mesosphere.com:8080/v2/tasks HTTP/1.1" 503 352 "-" "curl/7.29.0" 275 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-219016)
Nov 30 13:52:07 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:07,141] INFO 172.16.1.16 - - [30/Nov/2015:13:52:04 +0000] "GET //int.srv3.hw.ca1.mesosphere.com:8080/v2/tasks HTTP/1.1" 503 352 "-" "curl/7.29.0" 275 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-218989)
Nov 30 13:52:07 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:07,448] INFO 172.16.1.15 - - [30/Nov/2015:13:52:04 +0000] "GET //int.srv3.hw.ca1.mesosphere.com:8080/v2/tasks HTTP/1.1" 503 352 "-" "curl/7.29.0" 276 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-219012)
Nov 30 13:52:07 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:52:07,717] INFO 172.16.1.14 - - [30/Nov/2015:13:52:04 +0000] "GET //int.srv3.hw.ca1.mesosphere.com:8080/v2/tasks HTTP/1.1" 503 352 "-" "curl/7.29.0" 275 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-218990)
Nov 30 13:53:01 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:53:01,782] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-219000)
Nov 30 13:53:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:53:04,273] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-218954)
Nov 30 13:53:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:53:04,463] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-218987)
Nov 30 13:53:04 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:53:04,543] INFO 0:0:0:0:0:0:0:1 - - [30/Nov/2015:13:53:01 +0000] "GET //localhost:8080/v2/tasks HTTP/1.1" 503 352 "-" "python-requests/2.6.0 CPython/2.7.5 Linux/3.10.0-229.7.2.el7.x86_64" 276 (mesosphere.chaos.http.ChaosRequestLog$$EnhancerByGuice$$dac83aa5:qtp983834017-219000)
Nov 30 13:53:05 srv3.hw.ca1.mesosphere.com marathon[2332]: [2015-11-30 13:53:05,061] INFO Proxying request to GET http://srv2.hw.ca1.mesosphere.com:8080/v2/tasks from srv3.hw.ca1.mesosphere.com:8080 (mesosphere.marathon.api.JavaUrlConnectionRequestForwarder$:qtp983834017-219011)
```

### PR DESCRIPTION
**Pull request by gkleiman**

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/Main.scala [hunk 1]
```diff
@@ -27,7 +30,32 @@ class MarathonApp extends App {
     val client = new ZooKeeperClient(
       Amount.of(conf.zooKeeperSessionTimeout().toInt, Time.MILLISECONDS),
       conf.zooKeeperHostAddresses.asJavaCollection
-    )
+    ) {
+      override def shouldRetry(e: KeeperException): Boolean = {
+        log.error(s"Got ZooKeeper exception: $e")
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/Main.scala]: I would put the exception into an extra argument such that you can see stack trace.

File: src/main/scala/mesosphere/marathon/Main.scala [hunk 2]
```diff
@@ -27,7 +30,32 @@ class MarathonApp extends App {
...
+        log.error("Committing suicide to avoid invalidating ZooKeeper state")
+
+        val f = Future {
+          shutdownAndWait()
+          // scalastyle:off magic.number
+          Runtime.getRuntime.halt(1)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/Main.scala]: why not use exit in the good case?

Comment (kolloch) [inline src/main/scala/mesosphere/marathon/Main.scala]: Did you actually also investigate the deadlocks a bit?

File: src/main/scala/mesosphere/marathon/Main.scala [hunk 3]
```diff
@@ -27,7 +30,32 @@ class MarathonApp extends App {
     val client = new ZooKeeperClient(
       Amount.of(conf.zooKeeperSessionTimeout().toInt, Time.MILLISECONDS),
       conf.zooKeeperHostAddresses.asJavaCollection
-    )
+    ) {
+      override def shouldRetry(e: KeeperException): Boolean = {
+        log.error(s"Got ZooKeeper exception: $e")
+        log.error("Committing suicide to avoid invalidating ZooKeeper state")
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/Main.scala]: I would add a comment which explains the code.
* Why is it important to stop the process?
* What is the problem with `exit`?

File: src/main/scala/mesosphere/marathon/Main.scala [hunk 4]
```diff
@@ -27,7 +30,32 @@ class MarathonApp extends App {
...
+
+        val f = Future {
+          shutdownAndWait()
+          // scalastyle:off magic.number
+          Runtime.getRuntime.halt(1)
+          // scalastyle:on
+        }(scala.concurrent.ExecutionContext.global)
+
+        try {
+          Await.result(f, 5.seconds)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/Main.scala]: Configurable?

**Comment by kolloch** [comment]
Do you think that you can add an integration test for this? And a unit test? (yes, that's possible. you can mock the Runtime interface if you expose that as a dependency)

File: src/main/scala/mesosphere/marathon/Main.scala [hunk 5]
```diff
@@ -27,7 +30,31 @@ class MarathonApp extends App {
     val client = new ZooKeeperClient(
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/Main.scala]: Can you create a new class outside of Main?

**Comment by aquamatthias** [comment]
@gkleiman Think you have to change/ignore the  
LeaderIntegrationTest: abdicate if the zk connection is dropped

Your setup now forces Marathon to shutdown, not to abdicate.

**Comment by aquamatthias** [comment]
Thanks!

**Pull request merged**